### PR TITLE
Reentrant channel model, for real this time (+ lots of cleanup)

### LINKED ISSUE #100: Channel
Now that our LES Julia ocean model is emerging, it would be good to srategise about getting a channel going. Important for many reasons but especially for the eddy param problem and also for looking at axymmetric flows, But it seems to me an easy one: we just treat y (across channel) like z. Solver issue is the same etc, am I correct? 
Science goal would be to train a closure on the x-averaged PV flux.
Thoughts, comments?
John

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR introduces a new reentrant channel model type, the `ChannelModel`, which is periodic in x, has walls in the y-direction, and has the usual rigid lid at the top. It's just a regular `Model` but with no-flux boundary conditions in the y-direction on all fields.

The halo regions are filled in appropriately to apply the free-slip boundary conditions. This is nice as we get a channel without having to change the operators or time-stepping at all. @jm-c has pointed out that things could get complicated with this approach when we get to implementing e.g. positivity-preserving advection schemes with having to fill the halo regions very frequently, but we agreed to cross that bridge when we get there. At that point, masks may turn out to be a better approach.

It would be nice to have a solid physics test before merging this. Perhaps https://github.com/climate-machine/Oceananigans.jl/issues/100#issuecomment-483255199 would work, although we've had some issues setting up a flow in thermal wind balance (see PR #179). It might still work as an eyeball test.

I will release a new minor version once this PR is merged.

Resolves #100

Helps with #242 
Helps with #258

### CODE REVIEW

**Comment by glwagner** [comment]
The boundary condition implementation should dispatch on the type of the boundary conditions.

The corresponding functionality in `OceanTurb.jl` is here

https://github.com/glwagner/OceanTurb.jl/blob/694f54e799ccf4971f89a8d285e9a825f3b97734/src/boundary_conditions.jl#L36

We probably need a kernel to fill the halo regions.

I don't understand @jm-c's comment about "having to fill the halo regions very frequently". A few more details on this point would be helpful to facilitate a discussion about it.

**Comment by ali-ramadhan** [comment]
> The boundary condition implementation should dispatch on the type of the boundary conditions.

I agree that's a good approach. To do this, we'd have to merge PR #288. Do you have some time to look at it?

> We probably need a kernel to fill the halo regions.

We do have a couple but they need to be rewritten. I ended up switching to broadcasts but I think kernels would be more readable, especially as the halo filling will be different for velocities (u, w) and tracers (T, S) with v being a special case as the walls are in the y-dimension. I plan to switch back to kernels before merging this.

> I don't understand @jm-c's comment about "having to fill the halo regions very frequently". A few more details on this point would be helpful to facilitate a discussion about it.

I think it may have been specifically in reference to how the 3rd order positivity-preserving DST (+ flux limiter) advection scheme is implemented (prototyped in PR #183). Not sure if other advection schemes pose a similar issue, but I figure we cross that bridge when we get there.

**Comment by ali-ramadhan** [comment]
I think this PR is ready to be merged. It might be too big to ask others to review but @jm-c and @christophernhill have a pretty good idea of what changes were made to accommodate channels.

Some notes:

A new Poisson solver has been added that solves Poisson's equation with periodic boundary conditions in the x and staggered Neumann BCs in the y and z. This involves computing a 2D DCT on the GPU, which has no native DCT, so I implemented the fast 2D cosine transform from Makhoul (1980). Unfortunately the 2D FCT does not generalize from the 1D version, and extra steps are involved. In particular, as Fourier coefficients need to be indexed in multiple ways in some statements, we need a second buffer/storage array until I can figure out whether this is necessary. A couple of 1D masks are also needed. Anyways, so this PNN Poisson solver is a little involved, but is still decently fast.

We have more tests and more GPU tests in general, but this also means that the test suite takes even longer to run especially with GPUs. I've moved each test set to it's respective file so we can use the regression tests as a stand-alone 

In particular, this PR heavily refactors `poisson_solvers.jl`. It's more modular, concrete, and better documented I still need to describe the algorithms in the docs, especially important as it's not easy to figure this out and the only references are sparse papers (no code or implementations as far as I can tell). `time_steppers.jl` has also been cleaned up. Several kernels were converted to broadcasts. I also removed several bits of code that aren't used anymore.

I considered making separate PRs to clean things up but I had to refactor and clean up to implement stuff so I ended up doing it all in this PR.

The front-end has not changed at all but this PR ended up changing/cleaning a lot of the backend so I'd like to merge it ASAP before it starts to go out of date.

Currently it's twice as slow as running a doubly periodic model. Will have to profile to figure out where it's spending that extra time. For reference, a P100 is roughly 3x slower than a V100.
```
Nvidia P100 N=128^3
Periodic: 12 ms/iteration
Channel:  28 ms/iteration
```

**Comment by glwagner** [comment]
Do these changes mean that we need to rewrite the PR that makes things more concrete? It looks like there will be a lot of merge conflicts.

File: src/time_steppers.jl [hunk 1]
```diff
@@ -68,21 +68,37 @@ function time_step!(model::Model{A}, Nt, Δt) where A <: Architecture
     tb = (threads=(Tx, Ty), blocks=(Bx, By, Bz))
     FT = eltype(grid)
 
+    # Field tuples for fill_halo_regions.
```
Comment (glwagner) [inline src/time_steppers.jl]: What's going on here?

File: src/halo_regions.jl [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    # scalars operations on CuArrays right now which are extremely inefficient.
+    # Below we directly index the underlying Array or CuArray to avoid the issue of
+    # broadcasting over an OffsetArray{CuArray}.
+
+    # @views @inbounds @. f[1-Hx:0,     :, :] = f[Nx-Hx+1:Nx, :, :]
+    # @views @inbounds @. f[Nx+1:Nx+Hx, :, :] = f[1:Hx,       :, :]
+    # @views @inbounds @. f[:,     1-Hy:0, :] = f[:, Ny-Hy+1:Ny, :]
+    # @views @inbounds @. f[:, Ny+1:Ny+Hy, :] = f[:,       1:Hy, :]
+
+    if bcs.u.y.left == BoundaryCondition(Periodic, nothing)
```
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply src/halo_regions.jl]: Might be a good approach once things get more complicated but I didn't as I would have had to define another 5 functions to fill in y halos.

**Comment by ali-ramadhan** [comment]
> I'm a little worried about the increase in size of the already-quite-large time_step! function, but otherwise as far as I can tell things look ok.

I agree and at some point in this PR I switched to using broadcasts for the permutations which gets rid of all the `idct_permute!` kernels and keeps one `calculate_pressure_source_term` kernel, but I couldn't get it to perform well on the GPU so I scrapped it. Worth revisiting...

> One question: is it the name of the field that needs to be known for halo regions, or the type (FaceFieldZ, FaceFieldX, etc). If its the type, then this suggests two improvements:

Yeah it's kind of hacked together for now. When you want to fill halo regions you feed in a tuple like `(:u, bcs, u_data)` which fills the halos of `u_data` assuming it's a u-velocity field with boundary conditions `bcs`.

It kind of crap because for pressure halos I fill them like I would fill in the w-velocity field. This makes the code confusing I think. Might be good to attach boundary conditions to each field, so that all fields get assigned boundary conditions, not just `u, v, w, T, S`.

Would be great if we can can adapt fields to work well on GPUs. I think last time we tried we couldn't get them to perform for some reason. One of our oldest issues: https://github.com/climate-machine/Oceananigans.jl/issues/13

File: src/time_steppers.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/time_steppers.jl]: See comment below.

**Comment by glwagner** [comment]
> Might be good to attach boundary conditions to each field, so that all fields get assigned boundary conditions, not just u, v, w, T, S.

That's what I mean --- add `bcs` as a property to `Field`, in addition to `data` and `grid`.
